models/kernels/reservoir/Reservoir_Outlet.hpp

```diff
@@ -0,0 +1,107 @@
+#ifndef NGEN_RESERVOIR_OUTLET_H
+#define NGEN_RESERVOIR_OUTLET_H
+
+#include <cmath>
+#include "reservoir_parameters.h"
+#include "reservoir_state.h"
+
+/**
+ * @brief Base Single Reservior Outlet Class
+ *
+ * This class is for a single reservior outlet that holds the parameters a, b, and the activation threshold, which is
+ * the height in meters the bottom of the outlet is above the bottom of the reservoir. This class also contains a
+ * function to return the velocity in meters per second of the discharge through the outlet. This function will only
+ * run if the reservoir storage passed in is above the activation threshold.
+ *
+ * To modify the details of the calculations for discharge velocity, subclasses should typically override the
+ * calc_velocity_meters_per_second_local function, rather than the velocity_meters_per_second function, which works at a
+ * higher-level.
+ */
+class Reservoir_Outlet
+{
+public:
+
+    /**
+     * @brief Default Constructor building an empty Reservoir Outlet Object
+     */
+    Reservoir_Outlet();
+
+    /**
+     * @brief Parameterized Constuctor that builds a Reservoir Outlet object
+     * @param a outlet velocity calculation coefficient
+     * @param b outlet velocity calculation exponent
+     * @param activation_threshold_meters meters from the bottom of the reservoir to the bottom of the outlet
+     * @param max_velocity_meters_per_second max outlet velocity in meters per second
+     */
+    Reservoir_Outlet(double a, double b, double activation_threshold_meters, double max_velocity_meters_per_second);
+
+    /**
+     * Set the locally maintained velocity state variable value.
+     *
+     * @param velocity_meters_per_second The velocity value to set, in meters per second.
+     */
+    void adjust_velocity(double velocity_meters_per_second);
+
+    /**
+     * @brief Accessor to return activation_threshold_meters
+     * @return activation_threshold_meters meters from the bottom of the reservoir to the bottom of the outlet
+     */
+    double get_activation_threshold_meters() const;
```

Comment:
Should reservoir outlets be dynamic?  I.e. should we mark this interface `virtual` or keep it this way and get the compiled type function reference?  I'm not sure of a case off the top of my head for dynamic outlet/reservoir types, but it might be worth considering.

Reply:
Are reservoir outlet thresholds (i.e., in the real world) dynamic?  If so, I would agree this should be virtual.  We might also consider adding a protected setter for the member variable.

Reply:
I can do these if we decide to go this route.